internal/cluster/job.go

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package cluster
+
+import (
+	"context"
+	batchv1 "k8s.io/api/batch/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+// CleanJob cleans up a job in a given namespace
+func CleanJob(client crclient.Client, name string, namespace string) error {
+	job, err := GetJobInNamespace(client, name, namespace)
+	if err != nil {
+		return err
+	}
+	err = DeleteJob(client, job)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetJobInNamespace finds a job with a given name in a namespace
+func GetJobInNamespace(client crclient.Client, name string, namespace string) (*batchv1.Job, error){
+	job := &batchv1.Job{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, job)
+	if err != nil {
+		return job, err
+	}
+	return job, nil
+}
+
+// deleteJob deletes a given job and cleans up any pods associated with it
+func DeleteJob(client crclient.Client, job *batchv1.Job) error {
+	err := CleanupPods(client, job)
```

Comment:
Is it necessary to clean up pods on a job? I assumed the job would be the owner of its pods.

Reply:
It seems to linger around until its cleaned up even though the job doesn't exist anymore:
```
NAME                                          READY   STATUS      RESTARTS   AGE
che-plugin-registry-7c995f8845-n76d7          1/1     Running     0          2m41s
devworkspace-controller-d4f9dc86-qsg7l        1/1     Running     0          2m39s
devworkspace-webhook-server-9b7549977-9dlbh   1/1     Running     0          2m27s
webhook-server-gen-cert-nvzfs                 0/1     Completed   0          2m31s
```
```
[joshpinkney@localhost devworkspace-operator]$ kubectl get job
No resources found in devworkspace-controller namespace.
```

Reply:
As more I test I become more confusing how k8s works :-D
After I removed pods removing, I got job related pob left without ownerRef:
```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2020-10-01T09:09:32Z"
  generateName: webhook-server-gen-cert-
  labels:
    app.kubernetes.io/component: cert-generator
    app.kubernetes.io/name: webhook-server-gen-cert
    controller-uid: 2cc52032-5a5c-4ede-a9b7-ec8cc7099125
    job-name: webhook-server-gen-cert
  name: webhook-server-gen-cert-7gq47
  namespace: devworkspace-controller
  resourceVersion: "10595"
  selfLink: /api/v1/namespaces/devworkspace-controller/pods/webhook-server-gen-cert-7gq47
  uid: 4c6bba20-c069-4533-8026-0db2e2a1e4e4
```
Seems like k8s or k8s-go-client bug

Reply:
Alright! https://thenewstack.io/deletion-garbage-collection-kubernetes-objects/
Orphan removal is the default for crclient? Really?!!! We should be careful with that.
So, we should do background removal, in my proposal I still left warning and cleanning up if pods are still there.